src/query/frontend/semantic/required_privileges.cpp

```diff
@@ -80,8 +80,19 @@ class PrivilegeExtractor : public QueryVisitor<void>, public HierarchicalTreeVis
     AddPrivilege(AuthQuery::Privilege::CREATE);
     return false;
   }
-  bool PreVisit(CallProcedure & /*unused*/) override {
-    // TODO: Corresponding privilege
+  bool PreVisit(CallProcedure &procedure) override {
+    const auto add_privilege = [&](const auto &procedures, const auto privilege) {
+      if (std::any_of(procedures.begin(), procedures.end(),
+                      [&](const auto procedure_name) { return procedure_name == procedure.procedure_name_; })) {
+        AddPrivilege(privilege);
+      }
+    };
+    constexpr std::array<std::string_view, 2> module_read_procedures{"mg.get_module_file", "mg.get_module_files"};
+    add_privilege(module_read_procedures, AuthQuery::Privilege::MODULE_READ);
+
+    constexpr std::array<std::string_view, 3> module_write_procedures{"mg.create_module_file", "mg.update_module_file",
+                                                                      "mg.delete_module_file"};
+    add_privilege(module_write_procedures, AuthQuery::Privilege::MODULE_WRITE);
```

Comment:
What do you mean? Store it with the `Module` object?
Different procedures from the same module could have different Privileges as is the case here.
Also, it made more sense to me because of the separation of concerns. I know it's a bit weird to reference the name of procedures here, but you know if you want to change the privileges where to check.
Tell me if you think something different would make more sense! :smile:

Reply:
Yeah, maybe not the module, but the procedure could hold the required privileges. However I agree with separation of concerns, so that's is a bad approach.

My biggest concern is referencing the names as pure strings. Maybe we could expose those names as `constexpr std::string_view` (`get_module_file`, `delete_module_file`) and the builtin module name (`mg`) also. That way we can use those names and concatenate them everywhere in the codebase where we need them. Not the best, but I think it is less prone to bugs than pure strings.

Reply:
I see, I only have a small problem with where to put the constants. Do you have any suggestions?

Reply:
At first, I would say inside modules but since these are completely separate we could add an extra header for that. However, for this simple case, I would be personally fine to keep it as is unless we got something that is really better. Open for suggestions here as well :D